lib/ansible/modules/cloud/vmware/vmware_host_config_manager.py

```diff
@@ -98,18 +99,70 @@ def __init__(self, module):
+
+    @staticmethod
+    def is_boolean(value):
+        if str(value).lower() in ['true', 'on', 'yes', 'false', 'off', 'no']:
+            return True
+        return False
+
+    @staticmethod
+    def is_truthy(value):
+        if str(value).lower() in ['true', 'on', 'yes']:
+            return True
+        return False
+
     def set_host_configuration_facts(self):
         changed = False
         for host in self.hosts:
             option_manager = host.configManager.advancedOption
             host_facts = {}
             for option in option_manager.QueryOptions():
-                host_facts[option.key] = option.value
+                host_facts[option.key] = dict(value=option.value)
+
+            for s_option in option_manager.supportedOption:
+                host_facts[s_option.key].update(
+                    option_type=s_option.optionType,
+                )
+
             change_option_list = []
             for option_key, option_value in self.options.items():
-                if option_key in host_facts and option_value != host_facts[option_key]:
-                    change_option_list.append(vim.option.OptionValue(key=option_key, value=option_value))
-                    changed = True
+                if option_key in host_facts:
+                    # Make sure option_type is defined some values do not have
+                    # it defined and appear to be read only.
+                    if 'option_type' in host_facts[option_key]:
+                        # We handle all supported types here so we can give meaningful errors.
+                        option_type = host_facts[option_key]['option_type']
+                        if self.is_boolean(option_value) and isinstance(option_type, vim.option.BoolOption):
+                            option_value = self.is_truthy(option_value)
+                        elif (isinstance(option_value, integer_types) or self.is_integer(option_value))\
+                                and isinstance(option_type, vim.option.IntOption):
+                            option_value = VmomiSupport.vmodlTypes['int'](option_value)
+                        elif (isinstance(option_value, integer_types) or self.is_integer(option_value, 'long'))\
+                                and isinstance(option_type, vim.option.LongOption):
+                            option_value = VmomiSupport.vmodlTypes['long'](option_value)
+                        elif isinstance(option_value, float) and isinstance(option_type, vim.option.FloatOption):
+                            pass
+                        elif isinstance(option_value, string_types) and isinstance(option_type, (vim.option.StringOption, vim.option.ChoiceOption)):
+                            pass
+                        else:
+                            self.module.fail_json(msg="Provided value is of type %s."
+                                                      " Option %s expects: %s" % (type(option_value), option_key, type(option_type)))
```

Comment:
I was converting option_type to a string due to its contents being quite useful (min value, max value, default value, choices for choiceOptions, etc.)

If the formatting of the output was not acceptable we should still consider adding these other fields to the error message for the "vmodl.fault.InvalidArgument" exception handler below.  Let me know what you think.  I am willing to add this if you agree.

Reply:
I would rather add all other fields to the error message.

Reply:
Hmm, seems it posted a comment which I deleted from the review so this is in here twice.  Just to make sure I under stand you want to extract the values from the type objects and include them in the vmodl.fault.InvalidArgument exception handler below?